File: src/TopDown2dRenderer.ts
import { GameState, Position } from './CarGame.js';
import { Vehicle } from './Vehicle.js';
import { BaseRenderer } from './BaseRenderer.js';

export interface CanvasVehicle {
    x: number;
    y: number;
    steeringAngle: number;
    width: number;
    length: number;
    color: string;
    lighterColor: string;
}

// initialise the renderer
export class TopDown2dRenderer extends BaseRenderer {

    // Constants =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    readonly roadColor = 'gray';
    readonly roadMarkingsColor = '#ededed';

    readonly treeColor = '#535e3b';
    readonly backgroundColor = '#7a8a26';
    //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

    canvasCenterInWorldY: number = 0  // Y will adjust to the player's y position

    constructor(gameState: GameState, canvas: HTMLCanvasElement, gameOver: HTMLDivElement) {
        super(gameState, canvas, gameOver);
    }

    render(): void {
        if (this.gameState.gameOver) {
            this.endGame(this.gameState.score);
            return
        }

        this.canvas.clearCanvas(this.backgroundColor);

        this.updateCanvasCenterInWorld(this.gameState.player.y);
        this.drawRoad();

        // Draw vehicles on the road
        this.renderVehicles();

        // Draw trees off the road
        this.renderTrees();

        // Draw the car at it's position as a box
        this.drawCar(this.translatedVehicle(this.gameState.player));
        this.renderScore(this.gameState.score);

    }



    updateCanvasCenterInWorld(playerY: number): void {
        const yOffset = this.gameState.player.length; // yOffset is to adjust for the player being at the bottom centre of the screen
        this.canvasCenterInWorldY = playerY - yOffset
    }

    drawRoad(): void { //TBC - this could be simpler if we don't constantly redraw

        for (let i = 0; i < this.gameState.road.length - 1; i++) {

            const segmentStart: [Position, Position] = this.gameState.road[i]
            const segmentEnd: [Position, Position] = this.gameState.road[i + 1]

            this.drawRoadSegment(this.translateWorldSegmentToCanvas(segmentStart), this.translateWorldSegmentToCanvas(segmentEnd))

        }

    }


    drawRoadSegment(segmentStart: [Position, Position], segmentEnd: [Position, Position]): void {
        this.drawRoadSegmentRoad(segmentStart, segmentEnd);
        this.drawRoadSegmentMarkings(segmentStart, segmentEnd);
    }

    drawRoadSegmentRoad(segmentStart: [Position, Position], segmentEnd: [Position, Position]): void {



        // Draw the road segment as a polygon
        //   c          d
        //   se.0       se.1
        //   +-----------+
        //   |           |
        //   +-----------+
        //   ss.0       ss.1
        //   a          b
        this.canvas.drawQuadrilateral(segmentStart[0], segmentStart[1], segmentEnd[1], segmentEnd[0], this.roadColor);
    }


    drawRoadSegmentMarkings(segmentStart: [Position, Position], segmentEnd: [Position, Position]): void {

        // draw Road Segment Lanes
        this.drawRoadSegmentBoundaries(segmentStart, segmentEnd, 1 / 3, this.roadMarkingsColor, 0.75, true)

        // draw road boundary on the far side of the road
        this.drawRoadSegmentBoundaries(segmentStart, segmentEnd, 0.99, this.roadMarkingsColor, 0.25)

    }

    drawRoadSegmentBoundaries(
        segmentStart: [Position, Position],
        segmentEnd: [Position, Position],
        percentageOfSegment: number,
        color: string,
        strokeWidth: number,
        dashed: boolean = false): void {
        // Draw the road markings at thirds across the segment
        //   c                d
        //   se.0            se.1
        //   +-me1-------me2-+
        //   |               |
        //   +-ms1-------ms2-+
        //   ss.0           ss.1
        //   a               b


        const msPercetangeChunkX = ((segmentStart[1].x - segmentStart[0].x) * percentageOfSegment)
        const msPercetangeChunkY = ((segmentStart[1].y - segmentStart[0].y) * percentageOfSegment)
        const ms1: Position = {
            x: segmentStart[0].x + msPercetangeChunkX,
            y: segmentStart[0].y + msPercetangeChunkY
        }

        const ms2: Position = {
            x: segmentStart[1].x - msPercetangeChunkX,
            y: segmentStart[1].y - msPercetangeChunkY
        }


        const mePercetangeChunkX = ((segmentEnd[1].x - segmentEnd[0].x) * percentageOfSegment)
        const mePercetangeChunkY = ((segmentEnd[1].y - segmentEnd[0].y) * percentageOfSegment)
        const me1: Position = {
            x: segmentEnd[0].x + mePercetangeChunkX,
            y: segmentEnd[0].y + mePercetangeChunkY
        }
        const me2: Position = {
            x: segmentEnd[1].x - mePercetangeChunkX,
            y: segmentEnd[1].y - mePercetangeChunkY
        }

        //first line - at start + percentage of segment
        this.canvas.drawLine(ms1, me1, color, strokeWidth, dashed);

        //second line - at (end - percentage) of segment
        this.canvas.drawLine(ms2, me2, color, strokeWidth, dashed);
    }




    translateWorldSegmentToCanvas(worldSegment: [Position, Position]): [Position, Position] {
        return [
            this.translateWorldToCanvas(worldSegment[0]),
            this.translateWorldToCanvas(worldSegment[1])
        ];
    }

    translateLengthOnYAxisToCanvas(length: number): number {
        return this.translateWorldToCanvas({ x: 0, y: 0 }).y - this.translateWorldToCanvas({ x: 0, y: length }).y
    }

    translateLengthOnXAxisToCanvas(length: number): number {
        return this.translateWorldToCanvas({ x: length, y: 0 }).x - this.translateWorldToCanvas({ x: 0, y: 0 }).x
    }

    translateWorldToCanvas(worldPosition: Position): Position {
        return {
            // X=0 should be centered on the canvas
            // World X: -10 .. 10 => 0 .. 20 => 0..10 //assumption is that the world x is only between -10 and 10
            // Canvas X: 0..MapWidth
            x: (worldPosition.x + 10) / 20 * this.initialMapSize,

            // World Y: 0..10 (displayed on screen)
            // Canvas Y: 0..MapHeight
            // Y should be at the bottom of the canvas
            y: (this.initialMapSize - (((worldPosition.y - this.canvasCenterInWorldY) / 10) * this.initialMapSize)),
        };
    }

    // Translate the player to the canvas coordinates
    // This is because the player is in world coordinates, but the canvas is in canvas coordinates
    translatedVehicle(vehicle: Vehicle): CanvasVehicle {

        let translatedCenter = this.translateWorldToCanvas(vehicle)

        return {
            x: translatedCenter.x,
            y: translatedCenter.y,
            steeringAngle: vehicle.steeringAngle,
            width: this.translateLengthOnXAxisToCanvas(vehicle.width),
            length: this.translateLengthOnYAxisToCanvas(vehicle.length),
            color: vehicle.color,
            lighterColor: vehicle.lighterColor
        };

    }

    drawCar(canvasPlayer: CanvasVehicle): void {
        //Draw main body box
        this.canvas.drawRect(
            canvasPlayer.x,
            canvasPlayer.y,
            canvasPlayer.steeringAngle,
            canvasPlayer.width,
            canvasPlayer.length,
            canvasPlayer.color
        );

        // Draw the roof of the car
        const roofScale = 0.67;
        this.canvas.drawRect(
            canvasPlayer.x,
            canvasPlayer.y,
            canvasPlayer.steeringAngle,
            canvasPlayer.width * roofScale,
            canvasPlayer.length * roofScale,
            canvasPlayer.lighterColor
        );



        // Draw headlights positioned at the front of the car, rotated with the car
        const cos = Math.cos(canvasPlayer.steeringAngle * Math.PI / 180);
        const sin = Math.sin(canvasPlayer.steeringAngle * Math.PI / 180);

        // One headlight at -1, and one at +1 - so we get a left and right headlight
        for (let i = -1; i <= 1; i += 2) {

            // Headlight positions in car's local coordinate system (before rotation)
            // Front of car is at -length/2 in local Y, headlights are at ±width/4 in local X
            const headlightLocalX = i * canvasPlayer.width / 4;
            const headlightLocalY = - canvasPlayer.length / 2;

            // Rotate and translate headlight positions to canvas coordinates
            const headlightX = canvasPlayer.x + (headlightLocalX * cos) - (headlightLocalY * sin);
            const headlightY = canvasPlayer.y + (headlightLocalX * sin) + (headlightLocalY * cos);

            // Draw headlights rotated to face the car's direction
            this.canvas.drawRect(
                headlightX,
                headlightY,
                canvasPlayer.steeringAngle,
                canvasPlayer.width / 5,
                canvasPlayer.length / 10,
                'yellow'
            );
        }


    }


    renderTrees(): void {
        // Render all trees in the game state
        // let i = 0
        for (const tree of this.gameState.trees) {
            const canvasPos = this.translateWorldToCanvas({ x: tree.x, y: tree.y });

            // Tree foliage
            this.canvas.drawEllipse(
                canvasPos.x,
                canvasPos.y,
                this.translateLengthOnXAxisToCanvas(tree.radius),
                this.translateLengthOnYAxisToCanvas(tree.radius),
                this.treeColor // Forest green color
            );
        }
    }

    renderVehicles(): void {
        // Render all vehicles in the game state
        for (const vehicle of this.gameState.vehicles) {
            this.drawCar(this.translatedVehicle(vehicle));
        }
    }

}

File: swift/CarDriveMAIN.playground/Sources/Renderers/TopDown2dRenderer.swift
import SwiftUI

public struct CanvasVehicle {
    public var x: Double
    public var y: Double
    public var steeringAngle: Double
    public var width: Double
    public var length: Double
    public var color: Color
    public var lighterColor: Color
}

public class TopDown2dRenderer {

    static let scaleFactor: Double = 10
    static let htmlCanvasSize: Double = 200 * scaleFactor

    let roadColor = Color.gray
    let roadMarkingsColor = Color(red: 0.93, green: 0.93, blue: 0.93)
    let treeColor = Color(red: 0.33, green: 0.37, blue: 0.23)
    let backgroundColor = Color(red: 0.48, green: 0.54, blue: 0.15)

    var canvasCenterInWorldY: Double = 0
    // Actual pixel dimension of the canvas – updated each render call
    var dimension: Double = TopDown2dRenderer.htmlCanvasSize

    public init() {}

    public func render(canvas: CanvasDrawer, gameState: GameState) {
        dimension = canvas.size.width   // use the real canvas size for coordinate mapping
        canvas.clearCanvas(backgroundColor: backgroundColor)

        updateCanvasCenterInWorld(playerY: gameState.player.y)
        drawRoad(canvas: canvas, gameState: gameState)
        renderVehicles(canvas: canvas, gameState: gameState)
        renderTrees(canvas: canvas, gameState: gameState)

        let playerOnCanvas = translatedVehicle(gameState.player)
        drawCar(canvas: canvas, car: playerOnCanvas)

        renderScore(canvas: canvas, score: gameState.score)
    }

    // MARK: - Camera

    func updateCanvasCenterInWorld(playerY: Double) {
        // canvasCenterInWorldY = playerY - gameState.player.length
        // TBC - needs player length accessible
        canvasCenterInWorldY = playerY - 2.0
    }

    // MARK: - Road

    func drawRoad(canvas: CanvasDrawer, gameState: GameState) {
        guard gameState.road.count > 1 else { return }
        for i in 0..<(gameState.road.count - 1) {
            let segmentStart = gameState.road[i]
            let segmentEnd = gameState.road[i + 1]
            drawRoadSegment(
                canvas: canvas,
                start: translateWorldSegment(segmentStart),
                end: translateWorldSegment(segmentEnd)
            )
        }
    }

    func drawRoadSegment(canvas: CanvasDrawer, start: (CGPoint, CGPoint), end: (CGPoint, CGPoint)) {
        canvas.drawQuadrilateral(
            a: start.0, b: start.1, c: end.1, d: end.0,
            color: roadColor
        )
        drawRoadSegmentMarkings(canvas: canvas, start: start, end: end)
    }

    func drawRoadSegmentMarkings(canvas: CanvasDrawer, start: (CGPoint, CGPoint), end: (CGPoint, CGPoint)) {
        drawRoadSegmentBoundaries(canvas: canvas, start: start, end: end, percentage: 1.0 / 3.0, color: roadMarkingsColor, strokeWidth: 0.75, dashed: true)
        drawRoadSegmentBoundaries(canvas: canvas, start: start, end: end, percentage: 0.99, color: roadMarkingsColor, strokeWidth: 0.25, dashed: false)
    }

    func drawRoadSegmentBoundaries(canvas: CanvasDrawer, start: (CGPoint, CGPoint), end: (CGPoint, CGPoint), percentage: Double, color: Color, strokeWidth: Double, dashed: Bool = false) {
        let msChunkX = (start.1.x - start.0.x) * percentage
        let msChunkY = (start.1.y - start.0.y) * percentage
        let ms1 = CGPoint(x: start.0.x + msChunkX, y: start.0.y + msChunkY)
        let ms2 = CGPoint(x: start.1.x - msChunkX, y: start.1.y - msChunkY)

        let meChunkX = (end.1.x - end.0.x) * percentage
        let meChunkY = (end.1.y - end.0.y) * percentage
        let me1 = CGPoint(x: end.0.x + meChunkX, y: end.0.y + meChunkY)
        let me2 = CGPoint(x: end.1.x - meChunkX, y: end.1.y - meChunkY)

        canvas.drawLine(from: ms1, to: me1, color: color, strokeWidth: strokeWidth, dashed: dashed)
        canvas.drawLine(from: ms2, to: me2, color: color, strokeWidth: strokeWidth, dashed: dashed)
    }

    // MARK: - Vehicles

    func renderVehicles(canvas: CanvasDrawer, gameState: GameState) {
        for vehicle in gameState.vehicles {
            drawCar(canvas: canvas, car: translatedVehicle(vehicle))
        }
    }

    func drawCar(canvas: CanvasDrawer, car: CanvasVehicle) {
        canvas.drawRect(x: car.x, y: car.y, rotation: car.steeringAngle, width: car.width, height: car.length, color: car.color)

        let roofScale = 0.67
        canvas.drawRect(x: car.x, y: car.y, rotation: car.steeringAngle, width: car.width * roofScale, height: car.length * roofScale, color: car.lighterColor)

        let cosA = cos(car.steeringAngle * .pi / 180)
        let sinA = sin(car.steeringAngle * .pi / 180)

        for i in stride(from: -1, through: 1, by: 2) {
            let headlightLocalX = Double(i) * car.width / 4
            let headlightLocalY = -car.length / 2

            let headlightX = car.x + (headlightLocalX * cosA) - (headlightLocalY * sinA)
            let headlightY = car.y + (headlightLocalX * sinA) + (headlightLocalY * cosA)

            canvas.drawRect(x: headlightX, y: headlightY, rotation: car.steeringAngle, width: car.width / 5, height: car.length / 10, color: .yellow)
        }
    }

    // MARK: - Trees

    func renderTrees(canvas: CanvasDrawer, gameState: GameState) {
        for tree in gameState.trees {
            let canvasPos = translateWorldToCanvas(Position(x: tree.x, y: tree.y))
            canvas.drawEllipse(
                x: canvasPos.x,
                y: canvasPos.y,
                radiusX: translateLengthOnXAxis(tree.radius),
                radiusY: translateLengthOnYAxis(tree.radius),
                color: treeColor
            )
        }
    }

    // MARK: - Score

    func renderScore(canvas: CanvasDrawer, score: Double) {
        // Scale relative to actual canvas size
        let scale    = dimension / TopDown2dRenderer.htmlCanvasSize
        let fontSize = 12.0 * TopDown2dRenderer.scaleFactor * scale
        let xPos     = 10.0 * TopDown2dRenderer.scaleFactor * scale
        let yPos     = 15.0 * TopDown2dRenderer.scaleFactor * scale
        canvas.drawText(x: xPos, y: yPos, fontSize: fontSize, text: "Score: \(Int(score))", color: Color(red: 0.49, green: 1.0, blue: 0.08))
    }

    // MARK: - Coordinate translation (world -> canvas)

    func translateWorldToCanvas(_ worldPosition: Position) -> CGPoint {
        let xRange: Double = 15
        let yScale: Double = 45
        return CGPoint(
            x: (worldPosition.x + xRange) / (xRange * 2) * dimension,
            y: dimension - (((worldPosition.y - canvasCenterInWorldY) / yScale) * dimension)
        )
    }

    func translateWorldSegment(_ segment: (Position, Position)) -> (CGPoint, CGPoint) {
        return (translateWorldToCanvas(segment.0), translateWorldToCanvas(segment.1))
    }

    func translateLengthOnXAxis(_ length: Double) -> Double {
        let origin = translateWorldToCanvas(Position(x: 0, y: 0))
        let offset = translateWorldToCanvas(Position(x: length, y: 0))
        return offset.x - origin.x
    }

    func translateLengthOnYAxis(_ length: Double) -> Double {
        let origin = translateWorldToCanvas(Position(x: 0, y: 0))
        let offset = translateWorldToCanvas(Position(x: 0, y: length))
        return origin.y - offset.y
    }

    func translatedVehicle(_ vehicle: Vehicle) -> CanvasVehicle {
        let center = translateWorldToCanvas(Position(x: vehicle.x, y: vehicle.y))
        return CanvasVehicle(
            x: center.x,
            y: center.y,
            steeringAngle: vehicle.steeringAngle,
            width: translateLengthOnXAxis(vehicle.width),
            length: translateLengthOnYAxis(vehicle.length),
            color: vehicle.color,
            lighterColor: vehicle.lighterColor
        )
    }
}
